R/list-modify.R

```diff
@@ -5,48 +5,61 @@
 #' both lists `list_modify()` takes the value from `y`, and `list_merge()`
 #' concatenates the values together.
 #'
-#' `list_update()` is a helper optimised for interactively
-#' modifying an existing list: you can use a formula (a quosure) to replace
-#' using existing values.
+#' `list_update()` handles quosures that can refer to values existing
+#' within the input list.
 #'
-#' @param _x,x List to modify
+#' @param .x List to modify.
 #' @param ... New values of a list. Use `NULL` to remove values.
 #'   Use a formula to evaluate in the context of the list values.
-#' @param y Replacement list
+#'   These dots have [splicing semantics][rlang::dots_list].
 #' @export
 #' @examples
 #' x <- list(x = 1:10, y = 4, z = list(a = 1, b = 2))
 #' str(x)
 #'
 #' # Update values
-#' str(list_modify(x, list(a = 1)))
+#' str(list_modify(x, a = 1))
 #' # Replace values
-#' str(list_modify(x, list(z = 5)))
-#' str(list_modify(x, list(z = list(a = 1:5))))
+#' str(list_modify(x, z = 5))
+#' str(list_modify(x, z = list(a = 1:5)))
 #' # Remove values
-#' str(list_modify(x, list(z = NULL)))
+#' str(list_modify(x, z = NULL))
 #'
 #' # Combine values
-#' str(list_merge(x, list(x = 11, z = list(a = 2:5, c = 3))))
+#' str(list_merge(x, x = 11, z = list(a = 2:5, c = 3)))
 #'
-#' # list_update is useful for interactive tweaking, because it
-#' # uses ... instead of a separate list
-#' str(list_update(x, a = 1))
-#' str(list_update(x, z = 5))
-#' str(list_update(x, z = list(a = 1:5)))
-#' str(list_update(x, z = NULL))
+#'
+#' # All these functions take dots with splicing. Use !!! or UQS() to
+#' # splice a list of arguments:
+#' l <- list(new = 1, y = NULL, z = 5)
+#' str(list_modify(x, !!! l))
 #'
 #' # In list_update() you can also use quosures to compute new values
-#' list_update(x, z1 = rlang::quo(z[1]))
+#' list_update(x, z1 = rlang::quo(z[[1]]))
 #' list_update(x, z = rlang::quo(x + y))
-list_modify <- function(x, y) {
-  list_recurse(x, y, function(x, y) y)
+list_modify <- function(.x, ...) {
+  dots <- dots_list(...)
+  list_recurse(.x, dots, function(x, y) y)
 }
-
 #' @export
 #' @rdname list_modify
-list_merge <- function(x, y) {
-  list_recurse(x, y, c)
+list_merge <- function(.x, ...) {
+  dots <- dots_list(...)
+  list_recurse(.x, dots, c)
+}
+#' @export
+#' @rdname  list_modify
+list_update <- function(.x, ...) {
+  dots <- dots_list(...)
+
+  formulas <- map_lgl(dots, is_bare_formula, lhs = FALSE, scoped = TRUE)
```

Comment:
Are the other changes reasonable? e.g. taking dots with splicing. It seems like this should now be our default.